core/src/main/java/hudson/model/Queue.java

```diff
@@ -1442,7 +1467,9 @@ public void maintain() {
                     if (r != null) {
                         r.run();
                     } else {
-                        new BlockedItem(top).enter(this);
+                        new BuildableItem(top).enter(this);
+                        new FlyWeightItem(top).enter(this);
+                        updateSnapshot();
```

Comment:
In general, I think we need to have comments at these points in the code where there's state changes to the queue i.e. what's getting added, removed etc and what is expected to happen.